src/dp.rs

```diff
@@ -1,16 +1,72 @@
 // SPDX-License-Identifier: MPL-2.0
 
 //! Differential privacy (DP) primitives.
-use std::fmt::Debug;
+//!
+//! There are three main traits defined in this module:
+//!
+//!  - `DifferentialPrivacyBudget`: Implementors should be types of budgets for DP-mechanisms.
+//!    Examples: zCDP, ApproximateDP (Epsilon-Delta), PureDP
+//!
+//!  - `DifferentialPrivacyDistribution`: Distribution from which noise is sampled.
+//!    Examples: DiscreteGaussian, DiscreteLaplace
+//!
+//!  - `DifferentialPrivacyStrategy`: This is a combination of choices for budget and distribution.
+//!    Examples: zCDP-DiscreteGaussian, EpsilonDelta-DiscreteGaussian
+//!
+use num_bigint::{BigInt, BigUint, TryFromBigIntError};
+use num_rational::{BigRational, Ratio};
 
-/// Positive rational number to represent DP and noise distribution parameters in protocol messages
-/// and manipulate them without rounding errors.
+/// Errors propagated by methods in this module.
+#[derive(Debug, thiserror::Error)]
+pub enum DpError {
+    /// Tried to use an invalid float as privacy parameter.
+    #[error(
+        "DP error: input value was not a valid privacy parameter. \
+             It should to be a non-negative, finite float."
+    )]
+    InvalidFloat,
+
+    /// Tried to construct a rational number with zero denominator.
+    #[error("DP error: input denominator was zero.")]
+    ZeroDenominator,
+
+    /// Tried to convert BigInt into something incompatible.
+    #[error("DP error: {0}")]
+    BigIntConversion(#[from] TryFromBigIntError<BigInt>),
+}
+
+/// Positive finite precision rational number to represent DP and noise distribution parameters in
+/// protocol messages and manipulate them without rounding errors.
 #[derive(Clone, Debug)]
-pub struct Rational {
-    /// Numerator.
-    pub numerator: u32,
-    /// Denominator.
-    pub denominator: u32,
+pub struct Rational(Ratio<BigUint>);
+
+impl Rational {
+    /// Construct a [`Rational`] number from numerator `n` and denominator `d`. Errors if denominator is zero.
+    pub fn from_unsigned<T>(n: T, d: T) -> Result<Self, DpError>
+    where
+        T: Into<u128>,
+    {
+        // we don't want to expose BigUint in the public api, hence the Into<u128> bound
+        let d = d.into();
+        if d == 0 {
+            Err(DpError::ZeroDenominator)
+        } else {
+            Ok(Rational(Ratio::<BigUint>::new(n.into().into(), d.into())))
+        }
+    }
+}
+
+impl TryFrom<f32> for Rational {
+    type Error = DpError;
+    fn try_from(value: f32) -> Result<Self, DpError> {
```

Comment:
I was a bit worried by this function, because we don't want to introduce rounding errors that underestimate the real privacy loss. It turns out that we're good: the `from_float` function outputs a rational that exactly matches the `f32` input (https://docs.rs/num-rational/latest/src/num_rational/lib.rs.html#282-301). Might be worth keeping track of this, in case other people are worried too.

Reply:
Yes right. I added a note.